Update and add annotations to sshd config for servers

emkll opened the pull request:
## Status

Work in progress

## Description of Changes

Fixes #5660 

- Update supported algorthms
- Disable some agent forwarding and tunnelling options
- Annotate and reorder configuration for readability

Sources:
- https://github.com/dev-sec/ansible-ssh-hardening
- https://github.com/arthepsy/ssh-audit / https://github.com/jtesta/ssh-audit/

## Testing

- [ ] provided config is good / adheres to best practices
- [ ] provided config is adequately annotated
- [ ] provided config eliminates log message described in #5660 
- [ ] on an existing Xenial install the config is applied properly when an Ansible run occurs (I have tested this)
## Deployment

#### New installs

The sshd configuration will be applied on new installs via Ansible

#### Existing installs

Existing installs will not be updated unattended, but applied when an Ansible run occurs

## Checklist


- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [x] These changes do not require documentation

emkll committed:
Updates sshd config

- Update supported algorthms
- Disable some agent forwarding and tunnelling options
- Annotate and reorder configuration for readability

Sources:
- https://github.com/dev-sec/ansible-ssh-hardening
- https://github.com/arthepsy/ssh-audit

zenmonkeykstop commented:
One minor note, would it be worth making an accompanying docs change specifically in https://docs.securedrop.org/en/stable/servers.html?highlight=ssh-keygen to use ed25519 for new installs??

zenmonkeykstop commented:
Tested on prod VMs  (using existing instructions to generate an RSA SSH key on the workstation):
- [x] installation completes successfully
- [ ] on subsequent playbook runs, or on attempts to connect via `ssh <serverAlias>` SSH fails with a host key mismatch, as an ECDSA host key was set up initially but an ED25519 key is now present.

Removing the `known_hosts` entries for app and mon and accepting the new host keys clears the error but it might be good to handle it more transparently.

zenmonkeykstop commented on install_files/ansible-base/roles/restrict-direct-access/templates/sshd_config:
```diff
@@ -22,20 +33,34 @@ ChallengeResponseAuthentication no
+# Cipher selection
+
+Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes128-ctr
+HostKeyAlgorithms ssh-ed25519,rsa-sha2-256,rsa-sha2-512
```
It looks like this ordering forces a new host key to be generated when the new config is applied.  Since ECDSA is being removed altogether I don't see that there's a way around this in Xenial. It might make sense to have `tailsconfig` remove the old host key entries on the workstation if found, and alert the user that they can expect to see prompts on their first SSH/playbook run.

emkll committed:
Use ditribution-default host key algorithms

ECDSA will be used by defaut for the client to authenticate the host.

Tor Onion Services will also provide another layer of authentication,
when using ssh over Tor.

emkll commented:
> Since ECDSA is being removed altogether I don't see that there's a way around this in Xenial. It might make sense to have tailsconfig remove the old host key entries on the workstation if found, and alert the user that they can expect to see prompts on their first SSH/playbook run.

This might also be an issue in Focal, in a default configuration, the Host key is still ECDSA.


We could preserve the same host key preferences as the previous sshd_config: Given by default ssh is exposed Tor, which provides a second layer of authentication via authenticated onion services, or only exposed through a local network,  we could simply preserve ECDSA / distribution defaults for host keys.

I will look into  @zenmonkeykstop 's suggestion above, and use Ansible to
1. Delete the server fingerprint in known_hosts, and have a user re-verify the fingerprints or
2. Have Ansible edit known_hosts and add server fingerprints there

This might need to be done before the reboot action, so in the main Ansible playbook

conorsch commented:
> We could preserve the same host key preferences as the previous sshd_config

Preserving as-is seems the most straightforward. Certainly for Xenial, there's no pressing need to change the config. For Focal, ideally we'd have a stronger default, but the gains from v2 -> v3 onions are pretty great already, and v3 is required for Focal. So I _don't_ think we need to block on managing the host keys, as long as you're comfortable backing out on the defaults for now, @emkll. 

It does strike me as worthwhile to consider hostkey management outside the context of config updates: on first install, the same hostkey for the IPv4 address must be re-confirmed for the ssh onions, which we needn't prompt admins to configure. Still, I don't think that work needs to go into this PR. We can file a separate issue for it, and address if it proves to be a sticking point during testing the backup-and-restore flow as part of the LTS upgrade path.

emkll commented:
This is ready for re-review, however the docs update identified in https://github.com/freedomofpress/securedrop/pull/5666#issuecomment-744487280 is still pending

kushaldas commented on install_files/ansible-base/roles/restrict-direct-access/templates/sshd_config:
```diff
@@ -22,20 +34,33 @@ ChallengeResponseAuthentication no
+
+# Cipher selection
+
+Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes128-ctr
```
Do we really need `chacha20-poly1305@openssh.com`? 
`chacha20` is specially being used from the cheap phones which does not have good hardware. And our users will use Tails on proper hardware to ssh.

emkll committed:
Update sshd config based on feedback

- As pointed out by @kushaldas, chacha20-poly1305 mostly mobile-specific cipher, and while historically present in the sshd configuration for SecureDrop, is not necessary to support Debian-based ssh clients.

- The UsePrivilegeSeparation option has been deprecated in OpenSSH 7.5 [1]. UsePrivilegeSeparation has defaulted to 'sandbox' since 6.1 [2] and to 'yes' since 3.3 [3].

[1] https://www.openssh.com/txt/release-7.5
[2] https://www.openssh.com/txt/release-6.1
[3] https://www.openssh.com/txt/release-3.3

kushaldas reviewed (commented):
The current changes look nice to me, did not test it in a server yet.

kushaldas reviewed (approved):
- [x] provided config is good / adheres to best practices
- [x] provided config is adequately annotated
- [x] provided config eliminates log message described in #5660 
- [x] on an existing Xenial install the config is applied properly when an Ansible run occurs 


I tested manually with ssh over Tor on both Focal and Xenial. Works as expected. Approving. ~ Will merge after standup.~ @conorsch  do you think is it ready for merge?
